### PR TITLE
add readline lib for xdag-cli

### PR DESCRIPTION
**Pull request by reymondtu**
1. add readline lib for xdag command  
2. finish #82 last command feature
3.add command completion feature
4.add macos lib
5.add autoconf readline check

### CODE REVIEW

File: client/terminal.c [hunk 1]
```diff
@@ -92,6 +97,7 @@ int terminal(void)
 			write(sock, cmd2, strlen(cmd2));
 		}
 		write(sock, cmd, strlen(cmd) + 1);
+        add_history(cmd);
```
Comment (xrdavies) [inline client/terminal.c]: Please take care of code style.

File: client/commands.c [hunk 2]
```diff
@@ -3,6 +3,8 @@
 #include <math.h>
 #include <stdlib.h>
 #include <ctype.h>
+#include <readline/readline.h>
```
Comment (xrdavies) [inline client/commands.c]: Regarding new involved library, is there any replacement for Windows?

File: client/commands.c [hunk 3]
```diff
@@ -48,73 +50,221 @@ void processXferCommand(char *nextParam, FILE *out, int ispwd, uint32_t* pwd);
...
+    // readline lib
+    rl_readline_name = "xdag";
+    rl_attempted_completion_function = xdag_com_completion;
+    
+    for (;;) {
+        if(transportFlags & XDAG_DAEMON) {
+            sleep(100);
+        } else {
+            cmd = readline("xdag> ");
+            if (*cmd){
```
Comment (jonano614) [inline client/commands.c]: Shouldn't be if(cmd) here?

File: client/commands.c [hunk 4]
```diff
@@ -48,73 +50,221 @@ void processXferCommand(char *nextParam, FILE *out, int ispwd, uint32_t* pwd);
...
+int xdag_com_xfer(char *, FILE*);
+int xdag_com_stats(char *, FILE*);
+int xdag_com_state(char *, FILE*);
+int xdag_com_help(char *, FILE*);
+int xdag_com_run(char *, FILE*);
+int xdag_com_terminate(char *, FILE*);
+int xdag_com_exit(char *, FILE*);
+
+char* xdag_com_generator(const char*, int);
+char** xdag_com_completion(const char *, int, int);
```
Comment (jonano614) [inline client/commands.c]: Function xdag_com_completion is already declared in header file. Not necessary to declare it again.

File: client/commands.h [hunk 5]
```diff
@@ -22,6 +22,14 @@ extern double xdagGetHashRate(void);
 
 #define XFER_MAX_IN				11
 
+typedef int (*xdag_com_func_t)(char*, FILE *);
+
+typedef struct {
+    char *name;                  /* User printable name of the function. */
+    xdag_com_func_t func;        /* Function to call to do the job.      */
+    char *doc;                   /* Documentation for this function.     */
+} XDAG_COMMAND;
```
Comment (jonano614) [inline client/commands.h]: This type used only command.c. Move declaration there.

File: client/commands.c [hunk 6]
```diff
@@ -48,73 +50,221 @@ void processXferCommand(char *nextParam, FILE *out, int ispwd, uint32_t* pwd);
...
+    
+    // readline lib
+    rl_readline_name = "xdag";
+    rl_attempted_completion_function = xdag_com_completion;
+    
+    for (;;) {
+        if(transportFlags & XDAG_DAEMON) {
+            sleep(100);
+        } else {
+            cmd = readline("xdag> ");
```
Comment (jonano614) [inline client/commands.c]: Do you know if this function works with Windows? Could you please check and add precompilation directives if necessary?
like 
#if WIN32
printf("%s> ", g_progname);
fflush(stdout);
fgets(cmd, XDAG_COMMAND_MAX, stdin);
#else
cmd = readline("xdag> ");
#endif

File: client/commands.c [hunk 7]
```diff
@@ -48,73 +50,221 @@ void processXferCommand(char *nextParam, FILE *out, int ispwd, uint32_t* pwd);
...
+    while ((name = commands[list_index].name)) {
+        list_index++;
+        if (strncmp(name, text, len) == 0)
+            return (dupstr(name));
+    }
+    return ((char *)NULL);
+}
+
+XDAG_COMMAND* find_xdag_command(char *name) {
+    register int i;
```
Comment (jonano614) [inline client/commands.c]: Why do you need 'register' modifier here?

File: client/commands.c [hunk 8]
```diff
@@ -48,73 +50,221 @@ void processXferCommand(char *nextParam, FILE *out, int ispwd, uint32_t* pwd);
...
+
+char ** xdag_com_completion(const char *text, int start, int end) {
+    char **matches;
+    matches = (char **)NULL;
+    if (start == 0)
+        matches = rl_completion_matches(text, xdag_com_generator);
+    return (matches);
+}
+
+char* dupstr(char* s){
```
Comment (jonano614) [inline client/commands.c]: How about 'strdup' function?

File: client/commands.c [hunk 9]
```diff
@@ -48,73 +50,221 @@ void processXferCommand(char *nextParam, FILE *out, int ispwd, uint32_t* pwd);
...
+}
+
+int xdag_com_help(char *args, FILE* out) {
+    printHelp(out);
+    return 0;
+}
+
+char ** xdag_com_completion(const char *text, int start, int end) {
+    char **matches;
+    matches = (char **)NULL;
```
Comment (jonano614) [inline client/commands.c]: Feel free to combine declaration and assign `char **matches = (char **)NULL;`

File: client/commands.c [hunk 10]
```diff
@@ -48,73 +50,221 @@ void processXferCommand(char *nextParam, FILE *out, int ispwd, uint32_t* pwd);
...
-		return -1;
-	} else if(!strcmp(cmd, "xfer")) {
-		processXferCommand(nextParam, out, ispwd, pwd);
-	} else if(!strcmp(cmd, "lastblocks")) {
-		processLastBlocksCommand(nextParam, out);
-	} else {
-		fprintf(out, "Illegal command.\n");
-	}
-	return 0;
+    uint32_t pwd[4];
```
Comment (jonano614) [inline client/commands.c]: Use tabs for indentation

File: client/commands.c [hunk 11]
```diff
@@ -48,73 +50,221 @@ void processXferCommand(char *nextParam, FILE *out, int ispwd, uint32_t* pwd);
...
-				break;
-			}
-		}
-	}
+    char *cmd;
+    if(!(transportFlags & XDAG_DAEMON)) printf("Type command, help for example.\n");
+    
+    // readline lib
+    rl_readline_name = "xdag";
+    rl_attempted_completion_function = xdag_com_completion;
```
Comment (xrdavies) [inline client/commands.c]: xdag_com_completion mallocs memory, will it be freed in readline lib or manually?

File: client/commands.c (hunk 8 above)
Comment (reymondtu) [reply client/commands.c]: ok

File: client/commands.c (hunk 7 above)
Comment (reymondtu) [reply client/commands.c]: sorry, copy some code, i think register can delete

File: client/commands.c (hunk 3 above)
Comment (reymondtu) [reply client/commands.c]: This is a mistake

File: client/commands.c (hunk 9 above)
Comment (reymondtu) [reply client/commands.c]: ok

File: client/commands.c (hunk 11 above)
Comment (reymondtu) [reply client/commands.c]: i check readline lib complete.c :
1160   if (rl_attempted_completion_function)
1161     {
1162       matches = (*rl_attempted_completion_function) (text, start, end);
1163       if (RL_SIG_RECEIVED())
1164         {
1165           _rl_free_match_list (matches);
1166           matches = 0;
1167           RL_CHECK_SIGNALS ();
1168         }
1169 
1170       if (matches || rl_attempted_completion_over)
1171         {
1172           rl_attempted_completion_over = 0;
1173           return (matches);
1174         }
1175     }

_rl_free_match_list() while free malloc memory 

1897 void
1898 _rl_free_match_list (matches)
1899      char **matches;
1900 {
1901   register int i;
1902 
1903   if (matches == 0)
1904     return;
1905 
1906   for (i = 0; matches[i]; i++)
1907     xfree (matches[i]);
1908   xfree (matches);
1909 }

File: client/commands.h [hunk 12]
```diff
@@ -41,4 +49,8 @@ int xfer_callback(void *data, xdag_hash_t hash, xdag_amount_t amount, xdag_time_
 long double amount2xdags(xdag_amount_t amount);
 xdag_amount_t xdags2amount(const char *str);
 
+char* xdag_com_generator(const char*, int);
+char** xdag_com_completion(const char *, int, int);
+XDAG_COMMAND* find_xdag_command(char*);
```
[the remark this answers is not in the document]
Comment (reymondtu) [reply client/commands.h]: xdag_com_completion also use at terminal.c

File: client/commands.h [hunk 13]
```diff
@@ -41,4 +49,8 @@ int xfer_callback(void *data, xdag_hash_t hash, xdag_amount_t amount, xdag_time_
 long double amount2xdags(xdag_amount_t amount);
 xdag_amount_t xdags2amount(const char *str);
 
+char* xdag_com_generator(const char*, int);
```
[the remark this answers is not in the document]
Comment (reymondtu) [reply client/commands.h]: ok

File: client/commands.c (hunk 4 above)
Comment (reymondtu) [reply client/commands.c]: ok

File: client/commands.h (hunk 5 above)
Comment (reymondtu) [reply client/commands.h]: ok

File: client/commands.c (hunk 6 above)
Comment (reymondtu) [reply client/commands.c]: good suggest

File: client/commands.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (jonano614) [reply client/commands.h]: Yes, but I suggest to move find_xdag_command, not xdag_com_completion

**Pull request closed without merging**